Migrate zero trust device posture rule

SirCortly opened the pull request:
A couple notes: 
* `name` attribute went from optional in V4 to required in V5. We will check the state during config transforms and use that value if it contains one, but otherwise we will not migrate the name attribute and the user will have to manually add the required name to their HCL. 
* The `input` attribute block is computed in V4 and defaults to empty strings. In V5 it defaults to null for attributes that are undefined in the HCL. For now we are going to migrate any empty values to null. If this causes issues it should be discovered in E2E testing.

Commit:
migrate zero_trust_device_posture_rule from v4 to v5

vaishakdinesh commented on internal/resources/zero_trust_device_posture_rule/v4_to_v5.go:
```diff
@@ -0,0 +1,472 @@
+
+// V4ToV5Migrator handles migration of device posture rule resources from v4 to v5
+type V4ToV5Migrator struct {
+	typeUpdater *structural.ResourceTypeUpdater
```
Dont think this is being used anywhere? Probably not needed.

vaishakdinesh commented on internal/resources/zero_trust_device_posture_rule/v4_to_v5.go:
```diff
@@ -0,0 +1,472 @@
+
+// convertInputBlockToAttribute converts input block to attribute syntax
+// Also handles nested locations block conversion and removal of running field
+func (m *V4ToV5Migrator) convertInputBlockToAttribute(body *hclwrite.Body, inputBlock *hclwrite.Block) {
```
I think we have a structural transformation method in the `transform` pkg for this? If not. Does it make sense to add it there?

vaishakdinesh commented on internal/resources/zero_trust_device_posture_rule/v4_to_v5.go:
```diff
@@ -0,0 +1,472 @@
+}
+
+// convertMatchBlocksToArray converts match blocks to an array attribute
+func (m *V4ToV5Migrator) convertMatchBlocksToArray(body *hclwrite.Body, matchBlocks []*hclwrite.Block) {
```
Pretty sure we have a method in the transform pkg for this.

vaishakdinesh commented on internal/resources/zero_trust_device_posture_rule/v4_to_v5.go:
```diff
@@ -0,0 +1,472 @@
+		return result, nil
+	}
+
+	resources.ForEach(func(key, resource gjson.Result) bool {
```
The iteration would be taken care of by the caller of the Handle() method in the pipeline. Unless you need to do something specific for this resource?

vaishakdinesh commented on internal/resources/zero_trust_device_posture_rule/v4_to_v5.go:
```diff
@@ -0,0 +1,472 @@
+}
+
+// isEmptyValue checks if a gjson.Result value is considered "empty" (default/zero)
+func isEmptyValue(value gjson.Result) bool {
```
Probably good to export and make it re-usable? I think there's a `utils.go` in the transforms pkg?

kamarjago replied:
Thanks

tamas-jozsa commented:
Could you rebase + improve the comprehensiveness of the integration testdata using prompt:subtask_08? make sure all tests pass

vaishakdinesh commented on cmd/tf-migrate/main.go:
```diff
@@ -438,6 +446,7 @@ func processStateFile(log hclog.Logger, p *pipeline.Pipeline, cfg config, apiCli
 		TargetVersion: cfg.targetVersion,
 		Resources:     cfg.resourcesToMigrate,
 		APIClient:     apiClient,
+		AllAST:        parsedConfigs,
```
`nit` maybe rename this to CFGFiles. And also maybe AST to CFGFile? But not a big deal.

SirCortly commented:
I updated the migrator tool to pass the parsed config files down into the state transformations through the ctx. This was so we could check if empty values are defined in the HCL and if so, preserve the empty value instead of transforming to null. If the state is run on its own, we will have a warning in the diagnostics that an empty value was converted to null and will require an update in place if it is defined in the HCL. 

I also updated integration / e2e tests to make sure everything is comprehensive. E2e tests are running without drift.

musa-cf commented:
Migration tests PR: https://github.com/cloudflare/terraform-provider-cloudflare/pull/6441
